### ISSUE TITLE
Last two classes eval are always nan and zero

### ISSUE THREAD

**Issue by deepankkumar** [open-mmlab/mmsegmentation#1245]
I have 8 different classes, but while evaluating the model last two classes are always nan and zero, shown below

```
+--------------+-------+-------+
|    Class     |  IoU  |  Acc  |
+--------------+-------+-------+
|     wall     |  95.2 | 99.34 |
|     beam     | 88.77 | 99.13 |
|    column    | 90.66 | 99.74 |
| window frame | 88.01 | 97.01 |
| window pane  | 92.21 | 99.39 |
|   balcony    | 82.34 | 99.63 |
|     slab     |  0.0  |  nan  |
|    ignore    |  nan  |  nan  |
+--------------+-------+-------+
```

My custom data script is this:
`classes = ( 'wall', 'beam', 'column', 'window frame', 'window pane', 'balcony', 'slab','ignore')
palette = [[202,150,150], [198,186,100], [167,183,186], [255,255,133], 
           [192,192,206], [32,80,160], [193,134,1],[70,70,70]]`

`from mmseg.datasets.builder import DATASETS
from mmseg.datasets.custom import CustomDataset

@DATASETS.register_module()
class StandfordBackgroundDataset(CustomDataset):
  CLASSES = classes
  PALETTE = palette
  def __init__(self, split, **kwargs):
    super().__init__(img_suffix='.png', seg_map_suffix='.png', 
                     split=split, reduce_zero_label=False, **kwargs)
    assert osp.exists(self.img_dir) and self.split is not None`

Here is my config file: 

`Config:
norm_cfg = dict(type='BN', requires_grad=True)
model = dict(
    type='EncoderDecoder',
    pretrained='open-mmlab://msra/hrnetv2_w48',
    backbone=dict(
        type='HRNet',
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=False,
        extra=dict(
            stage1=dict(
                num_modules=1,
                num_branches=1,
                block='BOTTLENECK',
                num_blocks=(4, ),
                num_channels=(64, )),
            stage2=dict(
                num_modules=1,
                num_branches=2,
                block='BASIC',
                num_blocks=(4, 4),
                num_channels=(48, 96)),
            stage3=dict(
                num_modules=4,
                num_branches=3,
                block='BASIC',
                num_blocks=(4, 4, 4),
                num_channels=(48, 96, 192)),
            stage4=dict(
                num_modules=3,
                num_branches=4,
                block='BASIC',
                num_blocks=(4, 4, 4, 4),
                num_channels=(48, 96, 192, 384)))),
    decode_head=dict(
        type='FCNHead',
        in_channels=[48, 96, 192, 384],
        in_index=(0, 1, 2, 3),
        channels=720,
        input_transform='resize_concat',
        kernel_size=1,
        num_convs=1,
        concat_input=False,
        dropout_ratio=-1,
        num_classes=8,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0)),
    train_cfg=dict(),
    test_cfg=dict(mode='whole'))
dataset_type = 'StandfordBackgroundDataset'
data_root = ''
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
crop_size = (512, 512)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations'),
    dict(type='Resize', img_scale=(1920, 1080), ratio_range=(0.5, 2.0)),
    dict(type='RandomCrop', crop_size=(512, 512), cat_max_ratio=0.75),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='PhotoMetricDistortion'),
    dict(
        type='Normalize',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        to_rgb=True),
    dict(type='Pad', size=(512, 512), pad_val=0, seg_pad_val=255),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_semantic_seg'])
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1920, 1080),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img'])
        ])
]
data = dict(
    samples_per_gpu=8,
    workers_per_gpu=8,
    train=dict(
        type='StandfordBackgroundDataset',
        data_root='',
        img_dir='image',
        ann_dir='label/component',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(type='LoadAnnotations'),
            dict(
                type='Resize', img_scale=(1920, 1080), ratio_range=(0.5, 2.0)),
            dict(type='RandomCrop', crop_size=(512, 512), cat_max_ratio=0.75),
            dict(type='RandomFlip', flip_ratio=0.5),
            dict(type='PhotoMetricDistortion'),
            dict(
                type='Normalize',
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='Pad', size=(512, 512), pad_val=0, seg_pad_val=255),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img', 'gt_semantic_seg'])
        ],
        split='splits/train.txt'),
    val=dict(
        type='StandfordBackgroundDataset',
        data_root='',
        img_dir='image',
        ann_dir='label/component',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1920, 1080),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ],
        split='splits/val.txt'),
    test=dict(
        type='StandfordBackgroundDataset',
        data_root='',
        img_dir='image',
        ann_dir='label/component',
        pipeline=[
            dict(type='LoadImageFromFile'),
            dict(
                type='MultiScaleFlipAug',
                img_scale=(1920, 1080),
                flip=False,
                transforms=[
                    dict(type='Resize', keep_ratio=True),
                    dict(type='RandomFlip'),
                    dict(
                        type='Normalize',
                        mean=[123.675, 116.28, 103.53],
                        std=[58.395, 57.12, 57.375],
                        to_rgb=True),
                    dict(type='ImageToTensor', keys=['img']),
                    dict(type='Collect', keys=['img'])
                ])
        ],
        split='splits/val.txt'))
log_config = dict(
    interval=10, hooks=[dict(type='TextLoggerHook', by_epoch=False)])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = 'checkpoints/latest.pth'
resume_from = None
workflow = [('train', 1)]
cudnn_benchmark = True
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0005)
optimizer_config = dict()
lr_config = dict(policy='poly', power=0.9, min_lr=0.0001, by_epoch=False)
runner = dict(type='IterBasedRunner', max_iters=50)
checkpoint_config = dict(by_epoch=False, interval=50)
evaluation = dict(interval=50, metric='mIoU', pre_eval=True)
work_dir = 'checkpoints'
seed = 0
gpu_ids = range(0, 1)
`

**Comment by MengzhangLI** [comment]
`num_classes` = number of foreground classes + 1.

Also, you could check the ratio of `slab` and `ignore` to see whether they are both too small.

If the number of samples of them are too small, the training may be not very stable. You could also try more times.

Best,

**Comment by deepankkumar** [comment]
Thanks for the reply!!
Yes, for the slab class, there are very few samples the dataset, it might not be in test dataset.

for Ignore,
As far as I understand, 
I should set num_classes = 8
and remove ignore from classes. (i.e., `classes = ( 'wall', 'beam', 'column', 'window frame', 'window pane', 'balcony', 'slab')`)

right?

**Comment by MengzhangLI** [comment]
No, the background should also be added.

num_classes = len(classes).

For example, in DRIVE dataset, it has only one foreground class, but it classes number is 2:

https://github.com/open-mmlab/mmsegmentation/blob/cb1bf9f372cb6ba28a0b7581dd62ebd5c44dcfa8/mmseg/datasets/drive.py#L18

**Comment by deepankkumar** [comment]
Okay!! 
Thats what I was doing earlier, but still getting nan value for last class, i.e., ignore.

`classes = ( 'wall', 'beam', 'column', 'window frame', 'window pane', 'balcony', 'slab','ignore')
palette = [[202,150,150], [198,186,100], [167,183,186], [255,255,133], 
           [192,192,206], [32,80,160], [193,134,1],[70,70,70]]`

Do I need to add ignore_label or reduce_zero_label??

**Comment by MengzhangLI** [comment]
No, I am not sure whether it is caused by small samples. You can try training again.

**Comment by deepankkumar** [comment]
Even If I change the sequence of classes, I get nan for the last one. any idea?

**Comment by MengzhangLI** [comment]
Sorry for late reply. Did you solve this problem?

Sometimes `NaN` would be rasied by relatively small number in total labels.
